accepted/2020/serializer/WriteableDomAndDynamic.md

```diff
@@ -0,0 +1,577 @@
+
+There are implementations of `IDynamicMetaObjectProvider` including [DynamicObject](https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.dynamicobject?view=net-5.0) and [ExpandObject](https://docs.microsoft.com/en-us/dotnet/api/system.dynamic.expandoobject?view=netcore-3.1). However, both implementations are not ideal:
+- There are many public members used for wiring up dynamic support, but are of little to no value for consumers so they would be confusing when mixed in with other methods intended to be used for writable DOM support.
+- It is not possible to efficiently combine writable DOM classes that are not tied to `dynamic` and `System.Linq.Expressions`:
+  - If combined, there would need to be an assembly reference to the very large `System.Linq.Expressions.dll` event when `dynamic` features are not needed.
+  - If separate, the `dynamic` classes delegate to separate writable DOM classes, which doubles the object instances and increases the concept count around interop between the types returned from `dynamic` vs. the writable DOM classes.
+- `ExpandoObject` does not support case insensitivity and throws an exception when accessing missing properties (a `null` is desired instead).
+- `DynamicObject` prevents potential optimizations including property-lookup.
+
+Thus, the design presented here for `dynamic` assumes:
+- Implementation of `IDynamicMetaObjectProvider`.
+- Use of the writable DOM without requiring a reference to `System.Linq.Expressions.dll` (when `dynamic` support is not desired).
+- Efficient implementation of tying both `IDynamicMetaObjectProvider` and the writable DOM classes into a single instance. Newtonsoft also has a single object model with [JToken](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_Linq_JToken.htm) which implements `IDynamicMetaObjectProvider`.
+
+## Adding a System.Text.Json.Dynamic.dll (STJD) assembly
+Having `System.Text.Json.dll` directly reference `System.Linq.Expressions.dll` is not desired:
+- Implementations of Blazor and stand-alone apps that do not use `dynamic` features do not want to carry the large `System.Linq.Expressions.dll`.
+  - Referencing `System.Linq.Expressions.dll` will require ~5.5MB of disk size. `System.Linq.Expressions.dll` is around 5MB in size but since it also references to `System.Linq` (~400K) and `System.ObjectModel` (~80K) (which that are not referenced by STJ) it becomes ~5.5MB (crossgen'd size).
+  - STJ is considering removing its reference to using reflection emit, so referencing `System.Linq.Expressions.dll` would also require keeping the references to the various System.Reflection.* assemblies. However, since these are currently implemented in the runtime, it is not a factor today for disk size.
+
+- STJ is somewhat being pushed downstack with usage by other framework assemblies including the recent "BinaryData" work, so basic questions around the assembly dependency tree and potential cycles come into play.
```

Comment:
> STJ is somewhat being pushed downstack with usage by other framework assemblies including the recent "BinaryData" work

BinaryData is a high-level type.  It is not being included in netcoreapp.  It should have zero impact on layering within netcoreapp.  If it does, we need to revisit BinaryData.

Reply:
I'll re-word that as being a possibility.